### ISSUE TITLE
Gpu inquiry

### ISSUE THREAD

**Issue by gyupro** [id4thomas/llama2-koalpaca#1]
Hi, Thanks for sharing your training code.

I was wondering how much gpu memory you need to train Q-lora(4bit) Koalpaca training. Can you provide some detailed information you applied to the model?

For example, how long you had to train for a Epoch, what gpus you used and so on. 

Thanks in advance.

**Comment by id4thomas** [comment]
Hi, I haven’t got to training the model thoroughly, but these are the configurations that I’m currently using.

- src: eval with 13b, noeval with 70b
- GPU: NVIDIA A100 80GB * 2 (distributed data parallel)

using ddp needs modification to the current code which aren’t commited here.
gradient checkpointing was disabled for the details below due to issues with running it in DDP

These are the GPU memory related details:

**1. approximate GPU memory usage after loading in 4bit config (only pretrained model)**
- 13b: 8G
- 70b: 38G

**2. OOM after certain steps**
when training with the full koalpaca dataset, it would cause OOM after a certain amount of time due to samples that are long (see: koalpaca_data_stats.ipynb)

so I applied a cutoff using token size (only uses samples with < 1024 tokens) which is about 76.6% of the full dataset. (length after mapping to llama style input)

**These are the configurations that I'm currently using**
  | 13b | 70b
-- | -- | --
lora (r,alpha,dropout) | (16,32,0.05) | (16,32,0.05)
per_device_batch_size | 16 | 2
gradient_accumulation_steps | 4 | 8
epochs | 5 | 3
learning_rate | 2e-5 | 2e-5

I haven’t properly recorded the memory usage and training time, but **13b model took about 8 hours** (with eval) with the above settings and **70b took about 24 hours**.

Considering that training was ran on a 80G card*2, training 70b model would require more truncation of the training data to be feasible on a smaller memory card. (or other paging methods).

Feel free to ask any questions! I'll answer as much as I can

**Comment by gyupro** [comment]
Thank you for the detailed information! I'm new to NLP tasks, and your insights are incredibly helpful. I thought training the 70b model would require more than two A100 80GB GPUs! I plan to work on training llama2-koalpaca and will leave a comment if I have any further questions.
